### PR TITLE
Add debug wireframe to `ModelExperimental`

### LINKED ISSUE #9066: 3D tile wireframe display is tangled
There's some [assumptions in the code](https://github.com/CesiumGS/cesium/blob/1.71/Source/Scene/Model.js#L4449-L4450
) that triangle winding will support display as a LINES primitive which seem unlikely with real world data.  e.g. in [the sandcastle](https://sandcastle.cesium.com/?src=3D%20Tiles%20Inspector.html&label=3D%20Tiles)

image

image

### PR DESCRIPTION
**Pull request by j9liu**
Closes #9066. This PR adds the `debugWireframe` option to `ModelExperimental`. The current wireframe code in `Model` leads to tangled / incorrect wireframes, but `ModelExperimental` implements it properly.

| `Model` | `Model Experimental |
| -------- | ----------------------- |
| image | image |
| image | image |
| image | image |

Here's a modified version of the [Model Experimental sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=tVdtb9s2EP4rhPehzmDTsuMkrpcWS9I07dCgQZ22wKZhoKSzRZgiBZJynBX57ztSL5Zjd+taJK98ueM9d/fwjo6VNJasONyBJi+IhDtyAYYXGf3k17phJ/bzCyUt4xJ02OmRL6EkxKSqEMmZ5BmzMCVWF9AL5cPBL6EMZXkijYWKlzQutAZpb3kGaKM6/7dCcCZfoS6da5W9NWpyHAy77uSwMwpGw/5w2A+e346C6ehkenhCg+FxcDwenxwdjifHk/Hzw5Pfw04oS3uDAblWCQgTyti7lPkZmvvDnegBE2Jhbad4fAmBnDMhlJLoUbmNiiAgRpF5IWPLlSTdg1oXHfab3ky3XnNgKR3gz4xluQD0hw1K24PSSmVke0YXImrMuq/hUT12/rj/D363/PsI/mzJpYSEXKRMs9j6nDyhB9esRv8XDnegB/8D+SvEq8iFynINxqALHstTwffWNsZqd7hY3uoCebkQdv7drnyUgltyrtbfhf6rmPFAf3IzqFGS4JviqyQ8LRu8ifb4x7h8pVUhE/IJUh6LJ4NeWqmMbM9+iM9vsdQwGSORv5cI3wAej27sbE12Cfxvof/TF8q4KfgeABZIjyuBqFh85hrmmmVwKVkkIEH8TBhf1Vu6af6oU7wBlnC5uOE2Tj8oUTpV7V0zm1KrPqAEk6Y7nAQHHlBQ/i3Ld3nunK8hee3M32qUnSudbbpFs2QothQmSjn1utG5AuxNzCpddRCptE2r2ISdOzB20ywEVP0BZ02G2lkptOiRFPgitVXeSoi5MtwLN7gumLY4YvLQd7FXsNAApkprfzg6pMHJeHw8fF4laTymwVFweBIcVwulFTcusZGqGVMTo0c01zxDkyswVEOmVnCG8S1zm1X5+5o8S5IKR50Lp3C5zgFlsB0z4SFfIYe6DTcdo6bE+V+veDOYRs3X0z3ZSB8lv52TFrvryLW4ijxqzaqTL4XguVE8oZ+vZpNxS2APPRqub8ByfI4U2Q3Kihn/G58lw9Fks8vWbneG/MGdUYBfzd72BZhubgjdezWqy3XQTpyPFKaJJfc3GFhugNoUZHdTBLxEUwnqNxKeqhnSWi3PbBOynYB97cIxuWgHev+9Gx8d9P5Lpj/cEiq9iVylRFuzPAUNVKNsYcjPZFNnykEdg1oPnyVIwvJxiF54NjrqNkRDd/PpFjMb4Xe4RT9c3lye3fbqOHvG+38PztCMySRmxmL5xoNvlRIR09cgizLEpnkUIlfiJUFpEmNiLBDMCIm498kQJJFfcJkogdfFyJZH4vVKVFy460IXYC8FuOH5/Vu8Wp1KJuw4a5UnS4kPXlVYap3hbsOifSIsz8X9eYVlI9qrje/RcVFwQN5HBvTKMbGtF3b2c7Xjk0RNEZlY8whajFwxUUDDyGwP4V2FcUKbDLjvTq9zauy9gJd1Qn/lWY4119WOLrYxC9jGMOBmEOFTCyyNjal7EiE/1fH9siFchAFb+M48JXoRse541CP1b0AnG228HZh1jNqUjPN1azlSOgHdL1m6vfmwY5rLvLBtACvQlmNt6DPBF3KKxSRJBOxa7VtH3dGW5XorUtaqbHt317SvmrDl/Rw/X/XvfDOYoh8ieax+OmjH+zThK8KTF3s+npFYMGNwZ14IXwTDzsvTAcrvqArl68h7dFyweyeWDl++KxcppacDnO7XbKjfpP+0DKe9z8GBSiFeRu5JVJLHoXFPCZwn+LLpuxtYi7l3xl7e9krdj3niP2A+8xaetWzOUnVH7mqtOkwbvP8A) that allows you to toggle the wireframe.

This should also work for tilesets with `Cesium.ExperimentalFeatures.enableModelExperimental = true.` For example, here's the result of the changes in the [3D Tiles Inspector sandcastle (localhost)](http://localhost:8080/Apps/Sandcastle/index.html?src=3D%20Tiles%20Inspector.html&label=3D%20Tiles)

| With `Model` | With `ModelExperimental` |
| -------------- | ----------------------------- |
| image | image |

**Note:** this required changes to `GltfIndexBufferLoader` because implementing the wireframe required the index buffer to be available on the CPU. Index buffers are now loaded as typed arrays only in WebGL1 since it's not possible to get data from the buffer directly. With WebGL2, either buffers or typed arrays are fine.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @j9liu!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/model-experimental-debug-wireframe/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

File: Source/Scene/ModelExperimental/buildDrawCommands.js [hunk 1]
```diff
@@ -75,6 +85,12 @@ export default function buildDrawCommands(
     primitiveRenderResources.boundingSphere
   );
 
+  let count = primitiveRenderResources.count;
+  if (debugWireframe) {
+    primitiveType = PrimitiveType.LINES;
+    count *= 2;
+  }
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/buildDrawCommands.js:92]: Hm, this works if the primitive was `TRIANGLES`, but there are other cases that need to be handled differently:

* If the primitive was `POINTS`, should `debugWireframe` be ignored?
* `TRIANGLE_STRIP` and `TRIANGLE_FAN` have fewer indices, so it's not as simple as `count * 2`

File: Source/Scene/ModelExperimental/buildDrawCommands.js [hunk 2]
```diff
@@ -154,3 +170,112 @@ function deriveTranslucentCommand(command) {
...
+    typedArray: typedArray,
+    usage: BufferUsage.STATIC_DRAW,
+    indexDatatype: indexDatatype,
+  });
+}
+
+/**
+ * @private
+ */
+function createWireframeIndexBuffer(primitiveRenderResources, frameState) {
```
Comment (ptrgags) [inline Source/Scene/ModelExperimental/buildDrawCommands.js:204]: Considering my other comment about handling `TRIANGLE_STRIP`/`TRIANGLE_FAN`, maybe we need a helper module for computing indices?

File: Source/Scene/ModelExperimental/buildDrawCommands.js (hunk 1 above)
Comment (ptrgags) [reply Source/Scene/ModelExperimental/buildDrawCommands.js:92]: Looking at [This Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=zVhtb9s2EP4rnFE0MurISdqgm+0GW5xszZY0QW10H+JgpaVLzIYmNZJy4gb+7ztSb5TrtnlxgRlBRJF3zz13PB1PiqTQhswY3IAib4iAG9IHzdJp+MHNBaNG5O77UhjKBKhRo0XuRoIQPaGxvNEdYlQKrZFYNLsjkSGFcGtAxEEOlU1mNy8PhoyDPhI6gchIdcJumbCakWPCinlr/kTGwJFVDhqtBHDoThJBSgIRl9F1GKVKgTBDNoW6c3+mnFFxQA0EO692X+/u7ITbu69e/7LTdCAZFx2BgMq8u8VlDoYYSwGMJzzz+Lrw5CIYn3N7S7JZ+xN0Ch0yagwzCQxosaJAy1RF0CkmCIqFYRv/BnSacEDGtF0LZDtHKQft3HD4SUtRYi9aXyOhqNA8Rd+eSmSfmmgCcXH1gB9PqX2a0H9T+GHMMnjMwQdSPEm5YZuR5FKtl1vfQuoHsnkP6ATm4AfJ0ymsOZ/+ZmaS1YLcTGblgRRz99YbLEvNDYd0zB/KCAuIoSJ6MqcSpxqth1f7VDFMUWqYp7Y2jh74Awn25TSRmpmnJlqJ440exuRMMmH6XKZP3cMKyBu+/2P/0Xz60sXZ9NdQIVaSqxl4NM13Uk0p1z+CYA79aGrrKRgrqRVV5LHUDhSN5I8g5oC/RevC/dfAsfPBcywT7JBUxHCJvVncWtmXuS4lb3yuBfZFMjWhQVvXQdm1+H2PkZKPqW0HYxlhqRcmvAJzyMEO9+dHMTaFucyoYRWXsWmS8Pk+EzETV7qy0SqQm6sYWerWzulYg5rZwulrjhq5W2jRSep0rCPFxhBcpiKyZYwEIKxa3Cw2r2jcMk30Jxfo2uXF41jUY/8tNjKxV12yYZekaIjz7QryrW42q3RzXWaYKDZlhs1AhwqmcgalaDfPhwrzpyXQwq4H+mVXHeZ4GJQye7pVQptUicpSNqg0/IbYp0pjTI3isama7VrmgwnuquckVbxDcsJh8Ry1qvVswxzjw9sE0JQ9tHiW24XYopmNsg31qCIijednSk6ZLqVDMwHhbZMXpu8Eqmr6C/E8EJ+lnA5lUM2XDFr+nBeSt0gMH44zhpXoPRVXUFMmZKtVv9/cCZenTqiZhFN6G2xvbYVbZLN0eixT9+QNkgkoCBVaSnWLoFDTR/BuqsDZX1YEEiUx3oaBrjwPq0kvCDYHfeCldKx0muT583sJYnjY1cT4DP1NqvZXmzlferPzk21gl4O7etxc094hS7PO7Zi5RCzf2Pzf+ajx7C7jtSB7b8jPL+2LML4dW7hgI0kVFvoNG+bd5qhx0boHwlYNQUG8cU/N13VNqWwG3Vd5e6emPAfO5c29Ldd0OT6O9zZb04zmVHxD0z3dvvyYp1+xtDyzqN0vml6iLpYLBhaDyPYC/hGilFS1bDMTJW+Im+9WAP4RUh4Q4dLpkJfKWh3R51sX3lk7oSLmy59eBpECEIOERnA4w4L3NhMKim8bVMyozmzn+mjZHIkkNb85PzyH7OFhi2buU2b1EihWePvkrCh4+aLzcMWZlS9Xx0utXszfYbtkS0YuZs/SM38pKHYkU+MgrswE5Wv6YTadS15KRQL71YWh3FYXL71cr0tevGDebn1JZRn5nF10fWnJIeTyyitgH5/d+RqLDnl2t8KZwBdqLj4WALUDGv8tWl/b1eE8gfD48PfhP/3jo/5f/5v9LJzVE8x73Ejs4jPl73tzcnRwcHxY+dNoNXquRu8V4fmV4aueMvbgD7A/NoD9MTXYEI/T6BoreqR1EcFe21ftxWxGWPxmxfdIEnGqNa5cppwP2GcsHXu9Nsp/ocqlO3ZPZ6A4nVuxyfbecTYZhmGvjberNct+t/Sklz3sJMbmfnOMzS5K5Z1Mp/z41yqamyHcYqO+YV8m8JCYUaxnneVGvhTuU3dF+f5ESg2EklwESW447zLViozl3GM2Y4jBnbBhmkB0PZa3GB+fopuGuEPKfnqPDLJhzXNv/B8) the old Model ignored wireframe mode for `POINTS`, but `ModelExperimental` attempts to anyway.

The results do look cool though! reminds me of a yarn ball.
image

File: Specs/Scene/ModelExperimental/ModelExperimentalSpec.js [hunk 3]
```diff
@@ -116,6 +123,50 @@ describe(
       scene.backgroundColor = Color.BLACK;
     }
 
+    function verifyDebugWireframe(model, options) {
```
Comment (ptrgags) [inline Specs/Scene/ModelExperimental/ModelExperimentalSpec.js:126]: make sure to consider `POINTS`/`TRIANGLE_STRIP`/`TRIANGLE_FAN` in tests too.

**Review by ptrgags — COMMENTED** [review]
@j9liu for `gl.TRIANGLES`, this is working very nicely! I had some comments about making this more robust to handle triangle strips/fans

**Comment by j9liu** [comment]
@ptrgags updated, here's a [Sandcastle](http://localhost:8080/Apps/Sandcastle/index.html#c=vVdtb9s2EP4rhPehzmDLsuMkrpcWS9Ik69CgQe22wOZhoKSzRZgiBZJynBX57zuSkizXDtY2yPJii+SR99zdw7tTLIU2ZMXgDhR5RQTckQvQrMiCT26uPWvFbnwhhaFMgJq1OuTLTBCiU1nw5EywjBoYE6MK6MzEw8EvMzET/sQg5jJeBnGhFAgzZRmgjvL83wvOqHiDe4O5ktlbLUfHYb9tT561BuGg3+33u+HL6SAcD07GhydB2D8Oj4fDk6PD4eh4NHx5ePLHrDUTXl+vR25kAlzPROxMytwI1f1pT3SACTGwNmM8fqpQ9YIDmRjFcrTIL+NG4BCjyLwQsWFSkPZBtRcNdotOTbuas2B7kxxi3UNTaM+D6F1zM38naQKqV+lyqnoLPr3angoWKFtDsD/9o+rZ2ma/H9yq/3zMlCsq/h9DUNG2GTjxRCM8Jcg55VzKJ5sRBD38m9As5+Bs8VzoeS2lku0R4o9+HP5kyYSAhFykVNHYuDvyjBbc0Ar93zd0F3r4HcjfIF5JLmSWK9AaTXBYngu+07ZRVpnD+HKqCswTOyT6HlM+Cs4MOZfrH0L/KGY80J1cP1QoSfhN/pUCnpcNTkXz+WlcvlayEAn5BCmL+bNB91pKJdujJ/H5LaZ+KmIk8o8S4RvA49G1nq3B92XBv1zhiusC7ABgwXK4EoiKxWemYK5oBpeCRhwSxE+5dlW2sTfNv6rcvwFNmFjcMhOnHyT3RpVrN9SkgZEfUIIK3e6PwgMHKPSfvpz6c+dsDcmVVT9VKDuXKttU73pKB1jiKfdy8qrecw3YK1AjVVnRhVQmLX0za92BNpvizaGs1ziqI9SMSqF4h6TAFqkp4+Yh5lIzJ1zjuqDK4BMVh66reAMLBaDLsHb7g8MgPBkOj/svyyANh0F4FB6ehMflhNdinz02UjZHgY7RoiBXLEOVK9CBgkyu4Az962OblfF7TJ4mSYmjioXdcLnOAWWwPaLcQbYVt11z0zJqTKz91YxTg2FUbD3eE430q+A3Y9Jgd+W5BleRR41RefIl5yzXkiXB5+vJaNgQ2EOPmusbsAzbwyK7RVk+Yf9gm9gfjDardG1XJ8gfXBmE+FOvbV+A8eaGBHuvRnm5DpqBc57CMNHk/hYdyzQEJgXR3iQBJ1FngqpnxVMVRVrL5ZmpXbbjsMcuHLZETUfvv3fDo4POf8l0+1tC3prIZkrUNclTUBAolC00+Zls8ox/qHxQ7cO2BEnom3W0wrHRUrcmGpqbj7eYWQu/w6Xgw+Xt5dm0U/nZMd59PVhFEyqSmGqD6RsPnkrJI6puQBTexbpu0pEr8ZKgNIkxMAYIRoREzNmkCZLITdhIeOBVMjL+SLxeiYwLe12CBZhLDvbx/P4tXq1WKTNrWW2lJUuBLyCyMIGxits1i/aJ0Dzn9+cllo1op1K+Z4/1ggXyPtKgVpaJzX2z1n6utlyQAl1EOlYsggYjV5QXUDMy20N4m2Gs0CYC9rfVaZ1qc8/hdRXQX1mWY861uaONZcwAljF0uO5F2GqBCWKtq5pEyE+Vf79sCBehwxauMo+JWkS0PRx0SPUfBqPNbrwdGHX02pgM83VjOpIK3xy6nqXbiw87qpnIC9MEsAJlGOaGLuVsIcaYTJKEw67WrrHUHWxprpYiaYzMtld3VbusCVvWz/F9t3vnisEY7eDJ19tPe01/nyZsRVjyas/rMok51RpX5gV3SXDWen3aQ/mdrVy6PPIeDef03oql/dfv/GQQBKc9HO7fWVO/Dv+pd6e5z8GCSiFeRrYl8uSxaGwrgeMEO5uuvYGVmO0z9vK24/d+zBP3wv/CaXjR0DlJ5R25q3ZVbtrg/Rc) with `TRIANGLE_STRIP` and `TRIANGLE_FAN` models to demonstrate that they work

File: Source/Core/WireframeIndexGenerator.js [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    wireframeIndices[index++] = point2;
+    wireframeIndices[index++] = point2;
+    wireframeIndices[index++] = firstPoint;
+  }
+
+  return wireframeIndices;
+}
+
+/**
+ * Generates a wireframe index buffer for a primitive.
```
Comment (ptrgags) [inline Source/Core/WireframeIndexGenerator.js:162]: I would also add that this function both re-indexes existing indices, or it creates them from scratch if `originalIndices` is undefined.

File: Source/Core/WireframeIndexGenerator.js [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+import IndexDatatype from "./IndexDatatype.js";
+import PrimitiveType from "./PrimitiveType.js";
+
+/**
+ * Utility functions for generating indices for model wireframes. The indices are
+ * outputted as typed arrays, which can then be put into buffers for rendering.
+ *
+ * @private
+ */
+export default function WireframeIndexGenerator() {}
```
Comment (ptrgags) [inline Source/Core/WireframeIndexGenerator.js:11]: Hm, maybe this should be a namespace? see `MortornOrder.js` for an example

**Review by ptrgags — COMMENTED** [review]
@j9liu code updates look solid, and the new models are working properly.

One thing I noticed is that for some reason the `GroundVehicle` model doesn't seem to render properly when switched to wireframe mode. Any idea why this might be?

image

**Comment by j9liu** [comment]
Turns out on Linux, normal mapping doesn't work with lines because it handles the `dFdx`, `dFdy` functions differenly. I added a `define` to make sure the code doesn't try to use normal maps if it is rendered in wireframe. 

@ptrgags let me know if this solves the issue for you.

**Comment by ptrgags** [comment]
@j9liu oh wait I think I fetched without pulling :sweat_smile:, let me try again

File: Source/Shaders/ModelExperimental/MaterialStageFS.glsl [hunk 6]
```diff
@@ -23,7 +23,7 @@ vec3 computeNormal(ProcessedAttributes attributes)
     vec3 ng = attributes.normalEC;
 
     vec3 normal = ng;
-    #ifdef HAS_NORMAL_TEXTURE
+    #if defined HAS_NORMAL_TEXTURE && !defined USE_WIREFRAME
```
Comment (ptrgags) [inline Source/Shaders/ModelExperimental/MaterialStageFS.glsl:26]: For consistency with existing code, let's put parentheses, e.g. `defined(HAS_NORMAL_TEXTURE)`

**Review by ptrgags — COMMENTED** [review]
@j9liu sorry for the false alarm, the changes are working:

image

had one small comment before I merge

**Comment by ptrgags** [comment]
@j9liu looks good, I'll merge once CI passes!
